File: impl/ocaml/src/at_apply.ml
(* @apply expansion (SPEC §6.10). *)

open Ast
open Utils

(* Orders names so that every name comes after its dependencies; [on_cycle]
   raises for a cycle. *)
let topological_sort names (dependencies : (string * string list) list) on_cycle =
  let result = ref [] in
  let state = Hashtbl.create 16 in
  let rec visit name =
    match Hashtbl.find_opt state name with
    | Some 2 -> ()
    | Some _ -> on_cycle name
    | None ->
        Hashtbl.replace state name 1;
        let deps = match List.assoc_opt name dependencies with Some d -> d | None -> [] in
        let deps = List.sort compare (List.filter (fun d -> List.mem_assoc d dependencies) deps) in
        List.iter visit deps;
        Hashtbl.replace state name 2;
        result := name :: !result
  in
  List.iter visit names;
  List.rev !result

(* Builds the error message for an unknown @apply candidate. *)
let apply_error_message candidate (ds : Design_system.t) =
  let prefix = ds.theme.prefix in
  if prefix <> "" && not (has_prefix candidate (prefix ^ ":")) then
    "Cannot apply unknown utility class `" ^ candidate ^ "`. Did you forget the `" ^ prefix ^ ":` prefix?"
  else if Hashtbl.mem ds.invalid_candidates candidate then
    "Cannot apply utility class `" ^ candidate ^ "` because it is explicitly disabled by `@source not inline(...)`."
  else begin
    let parts = segment candidate ':' in
    let variants = List.filteri (fun i _ -> i < List.length parts - 1) parts in
    let variants = match variants with _ :: rest when prefix <> "" -> rest | v -> v in
    match List.find_opt (fun v -> Design_system.parse_variant ds v = None) variants with
    | Some v -> "Cannot apply unknown variant `" ^ v ^ "` in `" ^ candidate ^ "`."
    | None ->
        if Theme.size ds.theme = 0 then
          "Cannot apply unknown utility class `" ^ candidate
          ^ "`. The theme is empty; are you missing `@import \"twill\";` or `@reference \"twill\";`?"
        else "Cannot apply unknown utility class `" ^ candidate ^ "`."
  end

let expand_apply_in (nodes : nodes) (root : node option) (ds : Design_system.t) =
  let features = ref 0 in
  let path = match root with Some r -> [ r ] | None -> [] in
  ignore
    (walk_from nodes
       (fun n u ->
         match n with
         | At_rule at when at.name = "@utility" -> Skip
         | At_rule at when at.name = "@apply" -> (
             let parent = match u.parent with Some p -> Some p | None -> root in
             match parent with
             | None -> Continue (* A top-level @apply is left untouched. *)
             | Some (Context _) when Directives.is_top_level_path u.path -> Continue
             | Some _ ->
                 if !(at.at_nodes) <> [] then Twill_error.fail "`@apply` cannot have a body.";
                 List.iter
                   (function
                     | At_rule a when a.name = "@keyframes" -> Twill_error.fail "You cannot use `@apply` inside `@keyframes`."
                     | _ -> ())
                   u.path;
                 let candidates = fields at.params in
                 let mixins = List.length (List.filter (fun c -> has_prefix c "--") candidates) in
                 if mixins = List.length candidates && candidates <> [] then Continue
                 else begin
                   if mixins > 0 then
                     Twill_error.failf "You cannot mix CSS mixins with utility classes in `@apply %s`." at.params;
                   let invalid = ref None in
                   let compiled =
                     Compile_candidates.compile_candidates ~ignore_important:true
                       ~on_invalid:(fun c -> if !invalid = None then invalid := Some (apply_error_message c ds))
                       candidates ds
                   in
                   (match !invalid with Some message -> Twill_error.fail message | None -> ());
                   let replacement = List.concat_map (fun (r : rule) -> !(r.rule_nodes)) compiled in
                   features := !features lor Features.at_apply;
                   replace_with u replacement;
                   Continue
                 end)
         | _ -> Continue)
       root ctx_empty path);
  !features

(* Expands every @apply in place. @apply inside @utility bodies is expanded
   first, in dependency order. Returns the features. *)
let substitute (ast : nodes) (ds : Design_system.t) =
  let features = ref 0 in
  let utility_nodes = ref [] in
  ignore
    (walk ast (fun n _ ->
         match n with
         | At_rule at when at.name = "@utility" ->
             let root = trim_suffix (String.trim at.params) "-*" in
             utility_nodes := (root, at) :: List.remove_assoc root !utility_nodes;
             Skip
         | _ -> Continue));
  let utility_order = List.rev_map fst !utility_nodes in
  let utility_order = List.fold_left (fun acc r -> if List.mem r acc then acc else acc @ [ r ]) [] utility_order in
  if !utility_nodes <> [] then begin
    let dependencies =
      List.map
        (fun root ->
          let node = List.assoc root !utility_nodes in
          let deps = ref [] in
          ignore
            (walk node.at_nodes (fun child _ ->
                 (match child with
                 | At_rule at when at.name = "@apply" ->
                     List.iter
                       (fun candidate ->
                         List.iter
                           (fun (parsed : Candidate.candidate) ->
                             if parsed.ckind = Candidate.Arbitrary_candidate then ()
                             else if not (List.mem_assoc parsed.croot !utility_nodes) then ()
                             else if parsed.croot = root then
                               Twill_error.failf
                                 "You cannot `@apply` the `%s` utility here because it creates a circular dependency." candidate
                             else if not (List.mem parsed.croot !deps) then deps := !deps @ [ parsed.croot ])
                           (Design_system.parse_candidate ds candidate))
                       (fields at.params)
                 | _ -> ());
                 Continue));
          (root, !deps))
        utility_order
    in
    let sorted =
      topological_sort utility_order dependencies (fun cycle ->
          Twill_error.failf "You cannot `@apply` the `%s` utility here because it creates a circular dependency." cycle)
    in
    List.iter
      (fun root ->
        let node = List.assoc root !utility_nodes in
        features := !features lor expand_apply_in node.at_nodes (Some (At_rule node)) ds)
      sorted
  end;
  !features lor expand_apply_in ast None ds

File: impl/go/at_apply.go
package twill

import "strings"

// SubstituteAtApply expands every @apply in place (SPEC §6.10). @apply
// inside @utility bodies is expanded first, in dependency order.
func SubstituteAtApply(ast *[]Node, ds *DesignSystem) (Features, error) {
	var features Features

	utilityNodes := map[string]*AtRule{}
	var utilityOrder []string
	Walk(ast, func(n Node, _ *WalkUtils) WalkAction {
		if at, ok := n.(*AtRule); ok && at.Name == "@utility" {
			root := strings.TrimSuffix(strings.TrimSpace(at.Params), "-*")
			utilityNodes[root] = at
			utilityOrder = append(utilityOrder, root)
			return Skip
		}
		return Continue
	})

	if len(utilityNodes) > 0 {
		dependencies := map[string]map[string]bool{}
		var failure error
		for _, root := range utilityOrder {
			node := utilityNodes[root]
			deps := map[string]bool{}
			Walk(&node.Nodes, func(child Node, _ *WalkUtils) WalkAction {
				at, ok := child.(*AtRule)
				if !ok || at.Name != "@apply" {
					return Continue
				}
				for _, candidate := range strings.Fields(at.Params) {
					for _, parsed := range ds.ParseCandidate(candidate) {
						if parsed.Kind == CandidateArbitrary {
							continue
						}
						if _, ok := utilityNodes[parsed.Root]; !ok {
							continue
						}
						if parsed.Root == root {
							failure = Errorf("You cannot `@apply` the `%s` utility here because it creates a circular dependency.", candidate)
							return Stop
						}
						deps[parsed.Root] = true
					}
				}
				return Continue
			})
			if failure != nil {
				return 0, failure
			}
			dependencies[root] = deps
		}
		sorted, err := TopologicalSort(utilityOrder, dependencies, func(cycle string) error {
			return Errorf("You cannot `@apply` the `%s` utility here because it creates a circular dependency.", cycle)
		})
		if err != nil {
			return 0, err
		}
		for _, root := range sorted {
			node := utilityNodes[root]
			f, err := expandApplyIn(&node.Nodes, node, ds)
			if err != nil {
				return 0, err
			}
			features |= f
		}
	}

	f, err := expandApplyIn(ast, nil, ds)
	if err != nil {
		return 0, err
	}
	return features | f, nil
}

func expandApplyIn(nodes *[]Node, root Node, ds *DesignSystem) (Features, error) {
	var features Features
	var failure error
	var path []Node
	if root != nil {
		path = []Node{root}
	}
	WalkFrom(nodes, func(n Node, u *WalkUtils) WalkAction {
		at, ok := n.(*AtRule)
		if !ok {
			return Continue
		}
		if at.Name == "@utility" {
			return Skip
		}
		if at.Name != "@apply" {
			return Continue
		}
		parent := u.Parent
		if parent == nil {
			parent = root
		}
		// A top-level @apply is left untouched.
		if parent == nil {
			return Continue
		}
		if _, ok := parent.(*Context); ok && isTopLevelPath(u.Path) {
			return Continue
		}
		if len(at.Nodes) > 0 {
			failure = Errorf("`@apply` cannot have a body.")
			return Stop
		}
		for _, ancestor := range u.Path {
			if a, ok := ancestor.(*AtRule); ok && a.Name == "@keyframes" {
				failure = Errorf("You cannot use `@apply` inside `@keyframes`.")
				return Stop
			}
		}
		candidates := strings.Fields(at.Params)
		mixins := 0
		for _, c := range candidates {
			if strings.HasPrefix(c, "--") {
				mixins++
			}
		}
		if mixins == len(candidates) && len(candidates) > 0 {
			return Continue
		}
		if mixins > 0 {
			failure = Errorf("You cannot mix CSS mixins with utility classes in `@apply %s`.", at.Params)
			return Stop
		}
		var invalid error
		compiled := CompileCandidates(candidates, ds, CompileCandidatesOptions{
			IgnoreImportant: true,
			OnInvalidCandidate: func(candidate string) {
				if invalid == nil {
					invalid = Errorf("%s", ApplyErrorMessage(candidate, ds))
				}
			},
		})
		if invalid != nil {
			failure = invalid
			return Stop
		}
		var replacement []Node
		for _, rule := range compiled {
			replacement = append(replacement, rule.Nodes...)
		}
		features |= FeatureAtApply
		u.ReplaceWith(replacement...)
		return Continue
	}, root, ContextMap{}, path)
	return features, failure
}

// ApplyErrorMessage builds the error message for an unknown @apply candidate.
func ApplyErrorMessage(candidate string, ds *DesignSystem) string {
	prefix := ds.Theme.Prefix
	if prefix != "" && !strings.HasPrefix(candidate, prefix+":") {
		return "Cannot apply unknown utility class `" + candidate + "`. Did you forget the `" + prefix + ":` prefix?"
	}
	if ds.InvalidCandidates[candidate] {
		return "Cannot apply utility class `" + candidate + "` because it is explicitly disabled by `@source not inline(...)`."
	}
	parts := Segment(candidate, ':')
	variants := parts[:len(parts)-1]
	if prefix != "" && len(variants) > 0 {
		variants = variants[1:]
	}
	for _, v := range variants {
		if ds.ParseVariant(v) == nil {
			return "Cannot apply unknown variant `" + v + "` in `" + candidate + "`."
		}
	}
	if ds.Theme.Size() == 0 {
		return "Cannot apply unknown utility class `" + candidate + "`. The theme is empty; are you missing `@import \"twill\";` or `@reference \"twill\";`?"
	}
	return "Cannot apply unknown utility class `" + candidate + "`."
}

// TopologicalSort orders names so that every name comes after its
// dependencies. onCycle builds the error for a cycle.
func TopologicalSort(names []string, dependencies map[string]map[string]bool, onCycle func(string) error) ([]string, error) {
	var result []string
	state := map[string]int{}
	var failure error
	var visit func(name string)
	visit = func(name string) {
		if failure != nil {
			return
		}
		switch state[name] {
		case 2:
			return
		case 1:
			failure = onCycle(name)
			return
		}
		state[name] = 1
		var deps []string
		for dep := range dependencies[name] {
			if _, ok := dependencies[dep]; ok {
				deps = append(deps, dep)
			}
		}
		sortStrings(deps)
		for _, dep := range deps {
			visit(dep)
		}
		state[name] = 2
		result = append(result, name)
	}
	for _, name := range names {
		visit(name)
	}
	if failure != nil {
		return nil, failure
	}
	return result, nil
}

func sortStrings(s []string) {
	for i := 1; i < len(s); i++ {
		for j := i; j > 0 && s[j] < s[j-1]; j-- {
			s[j], s[j-1] = s[j-1], s[j]
		}
	}
}
